File: data_warehousing/connections/scripts/postgresql/PostgreSQLExample.java
import java.sql.*;
import java.util.Properties;

public class PostgreSQLExample {

    // Read from environment variables — never hardcode passwords in source code
    private static final String HOST     = System.getenv().getOrDefault("PG_HOST", "localhost");
    private static final String PORT     = System.getenv().getOrDefault("PG_PORT", "5432");
    private static final String DATABASE = System.getenv().getOrDefault("PG_DATABASE", "postgres");
    private static final String USER     = System.getenv().getOrDefault("PG_USER", "postgres");
    private static final String PASSWORD = System.getenv().getOrDefault("PG_PASSWORD", "");

    private static final String URL =
        String.format("jdbc:postgresql://%s:%s/%s", HOST, PORT, DATABASE);

    public static void main(String[] args) {
        Properties props = new Properties();
        props.setProperty("user", USER);
        props.setProperty("password", PASSWORD);
        props.setProperty("sslmode", "prefer");
        props.setProperty("connectTimeout", "10");

        try (Connection conn = DriverManager.getConnection(URL, props)) {
            conn.setAutoCommit(false);

            System.out.println("Connected to PostgreSQL!");

            // Create table
            try (Statement stmt = conn.createStatement()) {
                stmt.execute("""
                    CREATE TABLE IF NOT EXISTS students (
                        id    SERIAL PRIMARY KEY,
                        name  VARCHAR(100) NOT NULL,
                        grade INTEGER
                    )
                """);
            }

            // Insert using a prepared statement (prevents SQL injection)
            String insertSql = "INSERT INTO students (name, grade) VALUES (?, ?) RETURNING id";
            try (PreparedStatement ps = conn.prepareStatement(insertSql)) {
                ps.setString(1, "Bob");
                ps.setInt(2, 10);
                try (ResultSet rs = ps.executeQuery()) {
                    if (rs.next()) {
                        System.out.println("Inserted student with id=" + rs.getInt(1));
                    }
                }
            }

            // Query all rows
            try (Statement stmt = conn.createStatement();
                 ResultSet rs = stmt.executeQuery("SELECT id, name, grade FROM students ORDER BY id")) {
                while (rs.next()) {
                    System.out.printf("id=%d  name=%-20s  grade=%d%n",
                        rs.getInt("id"), rs.getString("name"), rs.getInt("grade"));
                }
            }

            conn.commit();

        } catch (SQLException e) {
            System.err.println("Database error: " + e.getMessage());
            e.printStackTrace();
        }
    }
}

File: data_warehousing/connections/scripts/postgresql/postgresql_example.py
import os
import psycopg2
from psycopg2 import sql, OperationalError

# Load credentials from environment variables — never hardcode passwords
HOST     = os.environ.get("PG_HOST", "localhost")
PORT     = int(os.environ.get("PG_PORT", 5432))
DATABASE = os.environ.get("PG_DATABASE", "postgres")
USER     = os.environ.get("PG_USER", "postgres")
PASSWORD = os.environ.get("PG_PASSWORD", "")

def connect():
    try:
        conn = psycopg2.connect(
            host=HOST,
            port=PORT,
            dbname=DATABASE,
            user=USER,
            password=PASSWORD,
            connect_timeout=10,
            sslmode="prefer",           # upgrade to SSL when the server supports it
        )
        conn.autocommit = False         # use explicit transactions
        return conn
    except OperationalError as e:
        print(f"Connection failed: {e}")
        raise


def main():
    conn = connect()
    try:
        with conn.cursor() as cur:
            # Read server version
            cur.execute("SELECT version();")
            row = cur.fetchone()
            print("Connected to:", row[0])

            # Create a table
            cur.execute("""
                CREATE TABLE IF NOT EXISTS students (
                    id      SERIAL PRIMARY KEY,
                    name    VARCHAR(100) NOT NULL,
                    grade   INTEGER
                );
            """)

            # Insert a row using parameterized query (prevents SQL injection)
            cur.execute(
                "INSERT INTO students (name, grade) VALUES (%s, %s) RETURNING id;",
                ("Alice", 9),
            )
            new_id = cur.fetchone()[0]
            print(f"Inserted student with id={new_id}")

            # Query rows
            cur.execute("SELECT id, name, grade FROM students ORDER BY id;")
            for record in cur.fetchall():
                print(record)

        conn.commit()
    except Exception as e:
        conn.rollback()
        print(f"Error: {e}")
        raise
    finally:
        conn.close()


if __name__ == "__main__":
    main()
